### PR TITLE
RN-587 Add pdf export server

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
RN-587

### Changes:
- New micro service `pdf-export-server` for pdf export, it will be used by `lesmis` and `tupaia`

### CODE REVIEW

File: packages/lesmis-server/src/routes/PDFExportRoute.ts [hunk 1]
```diff
@@ -25,67 +20,20 @@ export type PDFExportRequest = Request<
 >;
 
 export class PDFExportRoute extends Route<PDFExportRequest> {
+  private readonly apiClient: TupaiaApiClient;
```
Comment (IgorNadj) [inline packages/lesmis-server/src/routes/PDFExportRoute.ts:23]: This should really be provided by server-boilerplate. Lesmis-server is an Orchestrator Api https://github.com/beyondessential/tupaia/blob/dev/packages/lesmis-server/src/app/createApp.ts#L46

Server-boilerplate already gives us an api-client, but only for the other type of server.
- Microservice Api: https://github.com/beyondessential/tupaia/blob/dev/packages/server-boilerplate/src/microService/api/ApiBuilder.ts#L90 <-- present
- Orchestrator Api: https://github.com/beyondessential/tupaia/blob/dev/packages/server-boilerplate/src/orchestrator/api/ApiBuilder.ts <-- missing

Can you please add the same `attachApiClientToContext` method to orchestrator api and then have lesmis-server call it in createApp instead. Then here in this route you can change `this.apiClient.pdfExport()` to `this.req.ctx.services.pdfExport()`.

File: packages/pdf-export-server/.npmrc [hunk 2]
```diff
@@ -0,0 +1 @@
+scripts-prepend-node-path=true
```
Comment (IgorNadj) [inline packages/pdf-export-server/.npmrc:1]: I don't know if this is needed?

File: packages/pdf-export-server/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  "scripts": {
+    "build": "npm run --prefix ../../ package:build:ts",
+    "build-dev": "npm run build",
+    "lint": "yarn package:lint:ts",
+    "lint:fix": "yarn lint --fix",
+    "prestart": "npm run build",
+    "start": "node dist",
+    "start-dev": "yarn package:start:backend-start-dev 9993 -ts",
+    "start-verbose": "LOG_LEVEL=debug yarn start-dev",
+    "test": "yarn package:test:withdb --runInBand",
```
Comment (IgorNadj) [inline packages/pdf-export-server/package.json:26]:
```suggestion
    "test": "yarn package:test",
```

File: packages/pdf-export-server/src/@types/express/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/**
```
Comment (IgorNadj) [inline packages/pdf-export-server/src/@types/express/index.d.ts:1]: Do you need this file?

**Review by IgorNadj — APPROVED** [review]
Small comment about apiClient but apart from that looks good!

File: packages/pdf-export-server/.npmrc (hunk 2 above)
Comment (billli0) [reply packages/pdf-export-server/.npmrc:1]: I've seeen it in other packages, thought that's the default setting. Do you suggest we remove it?

Comment (IgorNadj) [reply packages/pdf-export-server/.npmrc:1]: Maybe something I added.... it's ok will clean up in a separate PR

File: packages/pdf-export-server/src/@types/express/index.d.ts (hunk 4 above)
Comment (billli0) [reply packages/pdf-export-server/src/@types/express/index.d.ts]: Nope. Have deleted it.

File: packages/server-boilerplate/src/orchestrator/api/ApiBuilder.ts [hunk 5]
```diff
@@ -155,14 +160,29 @@ export class ApiBuilder {
...
   public use<T extends ExpressRequest<T> = Request>(
     path: string,
-    middleware: RequestHandler<Params<T>, ResBody<T>, ReqBody<T>, Query<T>>,
+    ...middleware: RequestHandler<Params<T>, ResBody<T>, ReqBody<T>, Query<T>>[]
   ) {
     this.app.use(
       this.formatPath(path),
       this.attachSession as RequestHandler<Params<T>, ResBody<T>, ReqBody<T>, Query<T>>,
-      middleware,
+      ...middleware,
```
Comment (billli0) [inline packages/server-boilerplate/src/orchestrator/api/ApiBuilder.ts:185]: Be consistent with `use` in `microService`

File: packages/lesmis-server/src/app/createApp.ts [hunk 6]
```diff
@@ -46,6 +47,7 @@ export function createApp() {
   const app = new OrchestratorApiBuilder(new TupaiaDatabase(), 'lesmis')
     .useSessionModel(LesmisSessionModel)
     .useAttachSession(attachSession)
+    .attachApiClientToContext(authHandlerProvider) // after useAttachSession
```
Comment (billli0) [inline packages/lesmis-server/src/app/createApp.ts:50]: Not sure if this comment `after useAttachSession` is necessary.. But it took me a while to figure it out.

**Comment by IgorNadj** [comment]
Much cleaner, thanks for making those changes!

**Pull request merged**